moto 1.1.22 breaks dynamodb scan

Hello, I have an issue that started with moto 1.1.22. Reproduce with:

test_moto_failure.py
```from moto import mock_dynamodb2
import boto3

@mock_dynamodb2
def test_moto_problem():

    client = boto3.client('dynamodb',
                          aws_access_key_id="FakeKey",
                          aws_secret_access_key='FakeSecretKey',
                          aws_session_token='FakeSessionToken')

    table = client.create_table(
        TableName='table_name',
        KeySchema=[
            {
                'AttributeName': 'id',
                'KeyType': 'HASH'
            }
        ],
        AttributeDefinitions=[
            {
                'AttributeName': 'time-bucket',
                'AttributeType': 'N'
            }
        ],
        ProvisionedThroughput={
            'ReadCapacityUnits': 5,
            'WriteCapacityUnits': 5
        }
    )

    resp = client.scan(TableName='table_name',
                       Select='ALL_ATTRIBUTES',
                       FilterExpression='#tb >= :dt',
                       ExpressionAttributeNames={"#tb": "time-bucket"},
                       ExpressionAttributeValues={":dt": {"N": str(100)}
                                                  }
                       )
```

Environment:

```
$ pip freeze
asn1crypto==0.23.0
awscli==1.11.170
boto==2.48.0
boto3==1.4.7
botocore==1.7.28
certifi==2017.7.27.1
cffi==1.11.2
chardet==3.0.4
colorama==0.3.7
cookies==2.2.1
cryptography==2.1.1
daiquiri==1.3.0
dicttoxml==1.7.4
docker==2.5.1
docker-pycreds==0.2.1
docutils==0.14
idna==2.6
Jinja2==2.9.6
jmespath==0.9.3
MarkupSafe==1.0
mock==2.0.0
moto==1.1.22
numpy==1.13.1
pandas==0.20.3
pbr==3.1.1
py==1.4.34
pyaml==17.10.0
pyasn1==0.3.7
pycparser==2.18
pytest==3.2.1
python-dateutil==2.6.1
pytz==2017.2
PyYAML==3.12
requests==2.18.4
rsa==3.4.2
s3transfer==0.1.11
scikit-learn==0.19.0
scipy==0.19.1
six==1.10.0
urllib3==1.22
websocket-client==0.44.0
Werkzeug==0.12.2
xmltodict==0.11.0
```

`$ python -m pytest -k "test_moto_problem"`

Error:

```
======================================================================================================= test session starts =======================================================================================================
platform darwin -- Python 3.6.2, pytest-3.2.1, py-1.4.34, pluggy-0.4.0
rootdir: /Users/william/Documents/bde-deckard/score_server, inifile:
collected 9 items                                                                                                                                                                                                                  

tests/test_moto_failure.py F

============================================================================================================ FAILURES =============================================================================================================
________________________________________________________________________________________________________ test_moto_problem ________________________________________________________________________________________________________

    @mock_dynamodb2
    def test_moto_problem():
    
        client = boto3.client('dynamodb',
                              aws_access_key_id="FakeKey",
                              aws_secret_access_key='FakeSecretKey',
                              aws_session_token='FakeSessionToken')
    
        table = client.create_table(
            TableName='table_name',
            KeySchema=[
                {
                    'AttributeName': 'id',
                    'KeyType': 'HASH'
                }
            ],
            AttributeDefinitions=[
                {
                    'AttributeName': 'time-bucket',
                    'AttributeType': 'N'
                }
            ],
            ProvisionedThroughput={
                'ReadCapacityUnits': 5,
                'WriteCapacityUnits': 5
            }
        )
        epoch = "1000"
    
        resp = client.scan(TableName='table_name',
                           Select='ALL_ATTRIBUTES',
                           FilterExpression='#tb >= :dt',
                           ExpressionAttributeNames={"#tb": "time-bucket"},
>                          ExpressionAttributeValues={":dt": {"N": str(100)}
                                                      }
                           )

tests/test_moto_failure.py:37: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../anaconda/envs/deckard/lib/python3.6/site-packages/botocore/client.py:312: in _api_call
    return self._make_api_call(operation_name, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <botocore.client.DynamoDB object at 0x107b044a8>, operation_name = 'Scan'
api_params = {'ExpressionAttributeNames': {'#tb': 'time-bucket'}, 'ExpressionAttributeValues': {':dt': {'N': '100'}}, 'FilterExpression': '#tb >= :dt', 'Select': 'ALL_ATTRIBUTES', ...}

    def _make_api_call(self, operation_name, api_params):
        operation_model = self._service_model.operation_model(operation_name)
        request_context = {
            'client_region': self.meta.region_name,
            'client_config': self.meta.config,
            'has_streaming_input': operation_model.has_streaming_input,
            'auth_type': operation_model.auth_type,
        }
        request_dict = self._convert_to_request_dict(
            api_params, operation_model, context=request_context)
    
        handler, event_response = self.meta.events.emit_until_response(
            'before-call.{endpoint_prefix}.{operation_name}'.format(
                endpoint_prefix=self._service_model.endpoint_prefix,
                operation_name=operation_name),
            model=operation_model, params=request_dict,
            request_signer=self._request_signer, context=request_context)
    
        if event_response is not None:
            http, parsed_response = event_response
        else:
            http, parsed_response = self._endpoint.make_request(
                operation_model, request_dict)
    
        self.meta.events.emit(
            'after-call.{endpoint_prefix}.{operation_name}'.format(
                endpoint_prefix=self._service_model.endpoint_prefix,
                operation_name=operation_name),
            http_response=http, parsed=parsed_response,
            model=operation_model, context=request_context
        )
    
        if http.status_code >= 300:
            error_code = parsed_response.get("Error", {}).get("Code")
            error_class = self.exceptions.from_code(error_code)
>           raise error_class(parsed_response, operation_name)
E           botocore.exceptions.ClientError: An error occurred (InternalFailure) when calling the Scan operation: Internal error. replace() argument 2 must be str, not float

../../../anaconda/envs/deckard/lib/python3.6/site-packages/botocore/client.py:601: ClientError
```

The error `replace() argument 2 must be str, not float` is very odd because clearly I'm sending `str(100)` as my `ExpressionAttributeValues`.

When using `pip install moto==1.1.21` the test passes.

--- a/moto/dynamodb2/comparisons.py
+++ b/moto/dynamodb2/comparisons.py
@@ -71,6 +71,12 @@ def get_filter_expression(expr, names, values):
             values_map[key] = value['BOOL']
         elif 'S' in value:
             values_map[key] = value['S']
+        elif 'NS' in value:
+            values_map[key] = tuple(value['NS'])
+        elif 'SS' in value:
+            values_map[key] = tuple(value['SS'])
+        elif 'L' in value:
+            values_map[key] = tuple(value['L'])
         else:
             raise NotImplementedError()
 
@@ -136,13 +142,9 @@ def get_filter_expression(expr, names, values):
 
             next_token = six.next(token_iterator)
             while next_token != ')':
-                try:
-                    next_token = int(next_token)
-                except ValueError:
-                    try:
-                        next_token = float(next_token)
-                    except ValueError:
-                        pass
+                if next_token in values_map:
+                    next_token = values_map[next_token]
+
                 tuple_list.append(next_token)
                 next_token = six.next(token_iterator)
 
@@ -155,10 +157,14 @@ def get_filter_expression(expr, names, values):
                 tokens2.append(tuple(tuple_list))
         elif token == 'BETWEEN':
             field = tokens2.pop()
-            op1 = int(six.next(token_iterator))
+            # if values map contains a number, it would be a float
+            # so we need to int() it anyway
+            op1 = six.next(token_iterator)
+            op1 = int(values_map.get(op1, op1))
             and_op = six.next(token_iterator)
             assert and_op == 'AND'
-            op2 = int(six.next(token_iterator))
+            op2 = six.next(token_iterator)
+            op2 = int(values_map.get(op2, op2))
             tokens2.append(['between', field, op1, op2])
 
         elif is_function(token):

--- a/moto/dynamodb2/comparisons.py
+++ b/moto/dynamodb2/comparisons.py
@@ -63,7 +63,7 @@ def get_filter_expression(expr, names, values):
         expr = expr.replace(key, value)
     for key, value in values.items():
         if 'N' in value:
-            expr.replace(key, float(value['N']))
+            expr = expr.replace(key, value['N'])
         else:
             expr = expr.replace(key, value['S'])
 
@@ -176,7 +176,12 @@ def get_filter_expression(expr, names, values):
                     token = float(token)
                 except ValueError:
                     pass
-            tokens2.append(token)
+
+            # Need to join >= <= <>
+            if len(tokens2) > 0 and ((tokens2[-1] == '>' and token == '=') or (tokens2[-1] == '<' and token == '=') or (tokens2[-1] == '<' and token == '>')):
+                tokens2.append(tokens2.pop() + token)
+            else:
+                tokens2.append(token)
 
     # Start of the Shunting-Yard algorithm. <-- Proper beast algorithm!
     def is_number(val):

--- a/moto/dynamodb2/comparisons.py
+++ b/moto/dynamodb2/comparisons.py
@@ -61,15 +61,21 @@ def get_filter_expression(expr, names, values):
     # Do substitutions
     for key, value in names.items():
         expr = expr.replace(key, value)
+
+    # Store correct types of values for use later
+    values_map = {}
     for key, value in values.items():
         if 'N' in value:
-            expr = expr.replace(key, value['N'])
+            values_map[key] = float(value['N'])
+        elif 'BOOL' in value:
+            values_map[key] = value['BOOL']
+        elif 'S' in value:
+            values_map[key] = value['S']
         else:
-            expr = expr.replace(key, value['S'])
+            raise NotImplementedError()
 
     # Remove all spaces, tbf we could just skip them in the next step.
     # The number of known options is really small so we can do a fair bit of cheating
-    #expr = list(re.sub('\s', '', expr))  # 'Id>5ANDattribute_exists(test)ORNOTlength<6'
     expr = list(expr)
 
     # DodgyTokenisation stage 1
@@ -169,13 +175,9 @@ def get_filter_expression(expr, names, values):
             tokens2.append(function_list)
 
         else:
-            try:
-                token = int(token)
-            except ValueError:
-                try:
-                    token = float(token)
-                except ValueError:
-                    pass
+            # Convert tokens back to real types
+            if token in values_map:
+                token = values_map[token]
 
             # Need to join >= <= <>
             if len(tokens2) > 0 and ((tokens2[-1] == '>' and token == '=') or (tokens2[-1] == '<' and token == '=') or (tokens2[-1] == '<' and token == '>')):

--- a/tests/test_dynamodb2/test_dynamodb.py
+++ b/tests/test_dynamodb2/test_dynamodb.py
@@ -581,24 +581,24 @@ def test_filter_expression():
     row2 = moto.dynamodb2.models.Item(None, None, None, None, {'Id': {'N': '8'}, 'Subs': {'N': '10'}, 'Desc': {'S': 'A description'}, 'KV': {'SS': ['test3', 'test4']}})
 
     # AND test
-    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id > 5 AND Subs < 7', {}, {})
+    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id > :v0 AND Subs < :v1', {}, {':v0': {'N': 5}, ':v1': {'N': 7}})
     filter_expr.expr(row1).should.be(True)
     filter_expr.expr(row2).should.be(False)
 
     # OR test
-    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id = 5 OR Id=8', {}, {})
+    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id = :v0 OR Id=:v1', {}, {':v0': {'N': 5}, ':v1': {'N': 8}})
     filter_expr.expr(row1).should.be(True)
 
     # BETWEEN test
-    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id BETWEEN 5 AND 10', {}, {})
+    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id BETWEEN :v0 AND :v1', {}, {':v0': {'N': 5}, ':v1': {'N': 10}})
     filter_expr.expr(row1).should.be(True)
 
     # PAREN test
-    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id = 8 AND (Subs = 8 OR Subs = 5)', {}, {})
+    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id = :v0 AND (Subs = :v0 OR Subs = :v1)', {}, {':v0': {'N': 8}, ':v1': {'N': 5}})
     filter_expr.expr(row1).should.be(True)
 
     # IN test
-    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id IN (7,8, 9)', {}, {})
+    filter_expr = moto.dynamodb2.comparisons.get_filter_expression('Id IN :v0', {}, {':v0': {'NS': [7, 8, 9]}})
     filter_expr.expr(row1).should.be(True)
 
     # attribute function tests

--- a/tests/test_dynamodb2/test_dynamodb.py
+++ b/tests/test_dynamodb2/test_dynamodb.py
@@ -684,6 +684,33 @@ def test_scan_filter2():
     assert response['Count'] == 1
 
 
+@mock_dynamodb2
+def test_scan_filter3():
+    client = boto3.client('dynamodb', region_name='us-east-1')
+    dynamodb = boto3.resource('dynamodb', region_name='us-east-1')
+
+    # Create the DynamoDB table.
+    client.create_table(
+        TableName='test1',
+        AttributeDefinitions=[{'AttributeName': 'client', 'AttributeType': 'S'}, {'AttributeName': 'app', 'AttributeType': 'N'}],
+        KeySchema=[{'AttributeName': 'client', 'KeyType': 'HASH'}, {'AttributeName': 'app', 'KeyType': 'RANGE'}],
+        ProvisionedThroughput={'ReadCapacityUnits': 123, 'WriteCapacityUnits': 123}
+    )
+    client.put_item(
+        TableName='test1',
+        Item={
+            'client': {'S': 'client1'},
+            'app': {'N': '1'},
+            'active': {'BOOL': True}
+        }
+    )
+
+    table = dynamodb.Table('test1')
+    response = table.scan(
+        FilterExpression=Attr('active').eq(True)
+    )
+    assert response['Count'] == 1
+
 
 @mock_dynamodb2
 def test_bad_scan_filter():

--- a/tests/test_dynamodb2/test_dynamodb.py
+++ b/tests/test_dynamodb2/test_dynamodb.py
@@ -656,6 +656,36 @@ def test_scan_filter():
 
 
 @mock_dynamodb2
+def test_scan_filter2():
+    client = boto3.client('dynamodb', region_name='us-east-1')
+
+    # Create the DynamoDB table.
+    client.create_table(
+        TableName='test1',
+        AttributeDefinitions=[{'AttributeName': 'client', 'AttributeType': 'S'}, {'AttributeName': 'app', 'AttributeType': 'N'}],
+        KeySchema=[{'AttributeName': 'client', 'KeyType': 'HASH'}, {'AttributeName': 'app', 'KeyType': 'RANGE'}],
+        ProvisionedThroughput={'ReadCapacityUnits': 123, 'WriteCapacityUnits': 123}
+    )
+    client.put_item(
+        TableName='test1',
+        Item={
+            'client': {'S': 'client1'},
+            'app': {'N': '1'}
+        }
+    )
+
+    response = client.scan(
+        TableName='test1',
+        Select='ALL_ATTRIBUTES',
+        FilterExpression='#tb >= :dt',
+        ExpressionAttributeNames={"#tb": "app"},
+        ExpressionAttributeValues={":dt": {"N": str(1)}}
+    )
+    assert response['Count'] == 1
+
+
+
+@mock_dynamodb2
 def test_bad_scan_filter():
     client = boto3.client('dynamodb', region_name='us-east-1')
     dynamodb = boto3.resource('dynamodb', region_name='us-east-1')
@@ -678,7 +708,6 @@ def test_bad_scan_filter():
         err.response['Error']['Code'].should.equal('ValidationError')
     else:
         raise RuntimeError('Should of raised ResourceInUseException')
-
 
 
 @mock_dynamodb2

